File: iOS/Gallery/CustomTasks/VoiceAssistant/ChatHistoryStore.swift
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Port of customtasks/voiceassistant/ChatHistoryStore.kt

import Foundation

private let TAG = "AGChatHistoryStore"
/// Cap on stored messages per conversation to keep storage and LLM context bounded.
private let MAX_STORED_MESSAGES = 100

/// Persists the voice chat history on disk, one conversation per character,
/// as a small JSON file. Used both to restore on-screen messages and to seed
/// the LLM's context when a character is (re)selected.
final class ChatHistoryStore {

    private lazy var dir: URL = {
        let base = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let url = base.appendingPathComponent("voice_chat_history", isDirectory: true)
        try? FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        return url
    }()

    private func fileFor(_ conversationId: String) -> URL {
        let safe = conversationId.replacingOccurrences(
            of: "[^A-Za-z0-9_-]",
            with: "_",
            options: .regularExpression
        )
        return dir.appendingPathComponent("\(safe).json")
    }

    /// Loads the stored messages for `conversationId` (empty if none).
    func load(_ conversationId: String) -> [VoiceMessage] {
        guard !conversationId.isEmpty else { return [] }
        let file = fileFor(conversationId)
        guard FileManager.default.fileExists(atPath: file.path),
              let data = try? Data(contentsOf: file),
              let arr = try? JSONSerialization.jsonObject(with: data) as? [[String: String]]
        else { return [] }
        return arr.compactMap { obj -> VoiceMessage? in
            guard let r = obj["r"], let t = obj["t"] else { return nil }
            let role: VoiceMessage.Role = (r == "u") ? .user : .assistant
            return VoiceMessage(role: role, text: t)
        }
    }

    /// Saves `messages` for `conversationId` (keeps only the most recent ones).
    func save(_ conversationId: String, messages: [VoiceMessage]) {
        guard !conversationId.isEmpty else { return }
        let arr = messages.suffix(MAX_STORED_MESSAGES)
            .filter { !$0.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .map { ["r": ($0.role == .user ? "u" : "a"), "t": $0.text] }
        guard let data = try? JSONSerialization.data(withJSONObject: arr) else { return }
        try? data.write(to: fileFor(conversationId))
    }
}

File: Android/src/app/src/main/java/com/google/ai/edge/gallery/customtasks/voiceassistant/ChatHistoryStore.kt
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.ai.edge.gallery.customtasks.voiceassistant

import android.content.Context
import android.util.Log
import dagger.hilt.android.qualifiers.ApplicationContext
import java.io.File
import javax.inject.Inject
import javax.inject.Singleton
import org.json.JSONArray
import org.json.JSONObject

private const val TAG = "AGChatHistoryStore"
// Cap how many messages we keep per conversation so storage (and the restored LLM context) stay
// bounded.
private const val MAX_STORED_MESSAGES = 100

/**
 * Persists the voice chat history on disk, one conversation per character, as a small JSON file.
 *
 * Used both to restore the on-screen messages and to seed the LLM's context when a character is
 * (re)selected, so a conversation survives character switches and app restarts.
 */
@Singleton
class ChatHistoryStore @Inject constructor(@ApplicationContext private val context: Context) {

  private val dir: File by lazy {
    File(context.filesDir, "voice_chat_history").apply { mkdirs() }
  }

  private fun fileFor(conversationId: String): File {
    val safe = conversationId.replace(Regex("[^A-Za-z0-9_-]"), "_")
    return File(dir, "$safe.json")
  }

  /** Loads the stored messages for [conversationId] (empty if none). */
  fun load(conversationId: String): List<ChatMessage> {
    if (conversationId.isEmpty()) return emptyList()
    val file = fileFor(conversationId)
    if (!file.exists()) return emptyList()
    return try {
      val arr = JSONArray(file.readText())
      buildList {
        for (i in 0 until arr.length()) {
          val obj = arr.getJSONObject(i)
          val role =
            if (obj.getString("r") == "u") ChatMessage.Role.USER else ChatMessage.Role.ASSISTANT
          add(ChatMessage(role = role, text = obj.getString("t")))
        }
      }
    } catch (e: Exception) {
      Log.w(TAG, "Failed to load history for '$conversationId'", e)
      emptyList()
    }
  }

  /** Saves [messages] for [conversationId] (skips empty messages; keeps only the most recent ones). */
  fun save(conversationId: String, messages: List<ChatMessage>) {
    if (conversationId.isEmpty()) return
    try {
      val arr = JSONArray()
      for (message in messages.takeLast(MAX_STORED_MESSAGES)) {
        if (message.kind != ChatMessageKind.TEXT || message.text.isBlank()) continue
        arr.put(
          JSONObject()
            .put("r", if (message.role == ChatMessage.Role.USER) "u" else "a")
            .put("t", message.text)
        )
      }
      fileFor(conversationId).writeText(arr.toString())
    } catch (e: Exception) {
      Log.w(TAG, "Failed to save history for '$conversationId'", e)
    }
  }
}
